### PR TITLE
KAFKA-6104: Added unit tests for ClusterConnectionStates.

### PR DESCRIPTION
**Pull request by soenkeliebau**
Added a few unit tests for ClusterConnectionStates.

### CODE REVIEW

**Comment by tedyu** [comment]
Please fix indentation warnings.

File: clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
+package org.apache.kafka.clients;
```
Comment (tedyu) [inline clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:18]: Add license header

File: clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    public void testClusterConnectionStateChanges() {
+        assertTrue(connectionStates.canConnect(nodeId, time.milliseconds()));
+
+        // Start connecting to Node and check state
+        connectionStates.connecting(nodeId, time.milliseconds());
+        assertEquals("Expected connection state to be CONNECTING after initializing connection attempt.", connectionStates.connectionState(nodeId), ConnectionState.CONNECTING);
+        assertTrue("Expected connection to be connecting after initializing connection attempt.", connectionStates.isConnecting(nodeId));
+        assertFalse("Expected connection not to be ready for requests during an ongoing connection attempt.", connectionStates.isReady(nodeId));
+        assertFalse("Expected connection not to be blocked during an ongoing connection attempt", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertFalse("Expected connection to have no ready nodes.", connectionStates.hasReadyNodes());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java]: I think most (all?) of these assertions are very obvious and the long text is just repeating the code. Does the text add value of is it just noise? Also, the lines should not be longer than the GitHub review window.

File: clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        assertTrue("Expected connection to be blocked after disconnect event.", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertFalse("Expected connection not to still be in connecting state after disconnect event.", connectionStates.isConnecting(nodeId));
+        assertFalse("Expected connection to have no ready nodes after disconnect event.", connectionStates.hasReadyNodes());
+        assertFalse("Expected connection to be unable to connect immediately after disconnect event.", connectionStates.canConnect(nodeId, time.milliseconds()));
+
+        // After disconnecting we expect a backoff value equal to the reconnect.backoff.ms setting (plus minus 20% jitter)
+        long lowerLimit = Math.round(reconnectBackoffMsTest * (1 - reconnectBackoffJitter));
+        long upperLimit = Math.round(reconnectBackoffMsTest * (1 + reconnectBackoffJitter));
+        long currentBackoff = connectionStates.connectionDelay(nodeId, time.milliseconds());
+        assertTrue("Expected initial backoff after disconnect to be within 20% jitter of 'reconnect.backoff.ms' parameter.", lowerLimit <= currentBackoff && currentBackoff <= upperLimit);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java]: If you use the `assertEquals` that takes a `double`, you can pass a `delta` value, which makes the code a lot more concise.

File: clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        connectionStates.disconnected(nodeId, time.milliseconds());
+        connectionStates.remove(nodeId);
+        assertTrue("Expected being able to reconnect to node after removing it from the pool.", connectionStates.canConnect(nodeId, time.milliseconds()));
+        assertFalse("Expected connection to be unblocked after removing it from the pool.", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertEquals("Expected backoff delay to be zero after removing it from the pool.", connectionStates.connectionDelay(nodeId, time.milliseconds()), 0L);
+    }
+
+    @Test
+    public void testMaxReconnectBackoff() {
+        Long effectiveMaxReconnectBackoff = Math.round(reconnectBackoffMaxTest * (1 + reconnectBackoffJitter));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java]: Seems like this should be a `long`.

**Review by ijuma — COMMENTED** [review]
Thanks a lot for the tests! Just a few minor comments.

**Comment by soenkeliebau** [comment]
@ijuma I've addressed all your comments. I agree on the assert messages, I guess I was just trying to be holier than the pope here :)
I've also added a testcase with more than one node to check that nothing gets confused in that scenario.

**Comment by ijuma** [comment]
LGTM, merged to trunk and 1.0 branches after minor style fixes.

**Pull request closed without merging**